text/0000-stable-simd.md

```diff
@@ -0,0 +1,718 @@
+target-feature` flag to the compiler. This flag to the compiler accepts a
+similar set of strings to the ones specified above and is already "stable".
+
+## The `cfg_feature_enabled!` Macro
+
+One mode of operation with vendor intrinsics is to compile *part* of a program
+with certain CPU features enabled but not the entire program. This way a
+portable program can be compiled which runs across a broad range of hardware
+which can still benefit from optimized implementations for particular hardware
+at runtime.
+
+The crux of this support in libstd is this macro provided by libstd,
+`cfg_feature_enabled!`. The macro will accept one argument, a string literal.
+The string can be any feature passed to `#[target_feature(enable = ...)]` for
+the platform you're compiling for. Finally, the macro will resolve to a `bool`
+result.
+
+For example on x86 you could write:
+
+```rust
+if cfg_feature_enabled!("sse4.1") {
+    println!("this cpu has sse4.1 features enabled!");
+}
+```
+
+It would, however, be an error to write this on x86 cpus:
+
+```rust
+cfg_feature_enabled!("neon"); //~ ERROR: neon is an ARM feature, not x86
+cfg_feature_enabled!("foo"); //~ ERROR: unknown target feature for x86
+```
+
+The macro is intended to be implemented in the `std` crate (**not** `core`) and
+made available via the normal macro preludes. The implementation of this macro
+is expected to be what [`stdsimd`][stdsimd] does today, notably:
+
+* The first time the macro is invoked all the local CPU features will be
+  detected.
+* The detected features will then be cached globally (currently in a bitset) for
+  the rest of the execution of the program.
+* Further invocations of `cfg_feature_enabled!` are expected to be cheap runtime
+  dispatches. (aka load a value and check whether a bit is set)
+* Exception: in some cases the result of the macro is statically known: for
+  example, `cfg_feature_enabled!("sse2")` when the binary is being compiled
+  with "sse42" globally. In these cases, none of the steps above are
+  performed and the macro just expands to `true`.
+
+The exact method of CPU feature detection various by platform, OS, and
+architecture. For example on x86 we make heavy use of the `cpuid` instruction
+whereas on ARM the implementation currently uses getauxval/`/proc` mounted
+information on Linux. It's expected that the detection will vary for each
+particular target, as necessary.
+
+Note that the implementation details of the macro today prevent it from being
+located in libcore. If getauxval or `/proc` is used that requires libc to be
+available or `File` in one form or another. These concerns are currently
+std-only (not available in libcore). This is also a conservative route for x86
+where it is possible to do CPU feature detection in libcore (as it's just the
+`cpuid` instruction), but for consistency across platforms the macro will only
+be available in libstd for now. This placement can of course be relaxed in the
+future if necessary.
```

Comment:
Code that is being compiled for `#[no_std]` on a platform with no obvious way to perform feature detection could just statically compile these macros to `false`, which would also help with portability. A lint could be added that checks for this scenario and defaults to `allow`, but if a user values SIMD, they could change the lint to `warn` or `error` so that they know when they're compiling for a target that doesn't current support SIMD feature detection in `#[no_std]`.

Those are my 2 cents, at least.

Reply:
Hm a good point yeah! I think I'd personally still be inclined though to take the more conservative route and only expose it in libstd for now. I could imagine that it may be surprising if your crate is fast with `extern crate std` and slow without it otherwise (due to lacking feature detection). In that sense I don't think it's a all-around-win to have a noop macro in libcore per se.